streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1097,13 +1102,17 @@ private int getNumStreamThreads(final boolean hasGlobalTopology) {
         return Optional.empty();
     }
 
-    // Returns the number of threads that are not in the DEAD state -- use this over threads.size()
+    // Returns the number of threads that are not in the DEAD or PENDING_SHUTDOWN state -- use this over threads.size()
     private int getNumLiveStreamThreads() {
         final AtomicInteger numLiveThreads = new AtomicInteger(0);
         synchronized (threads) {
             processStreamThread(thread -> {
                 if (thread.state() == StreamThread.State.DEAD) {
+                    log.debug("Trimming thread {} from the threads list since it's state is {}", thread.getName(), StreamThread.State.DEAD);
                     threads.remove(thread);
+                } else if (thread.state() == StreamThread.State.PENDING_SHUTDOWN) {
```

Comment:
are we risking a memory overflow with this?

Reply:
Yep -- but imo this should be pretty rare. To hit an OOM you would need to have a relatively huge cache, and few total threads (so that the cache size per thread is still large), and then you'd need the new thread to start up, rebalance to get tasks, finish initializing and restoring them and then start consuming enough data to fill up the cache all before the old thread gets around to closing it's state stores. 
Sure, threads can sometimes hang during termination but in my experience that's usually after the store closure (and hopefully hanging is itself rare).

My point is, this should be good enough for now, and I'd like to stick with the simplest solution this late in the game.

Reply:
You can file an improvement ticket and list some ideas for how to handle this (eg scheduler that periodically skims the `threads` list and resizes the cache when it finds any DEAD threads to trim) but I'm optimistic this won't be an issue for now